cmd/metrics-v3-cluster-usage.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2015-2024 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/minio/minio/internal/logger"
+)
+
+const (
+	usageSinceLastUpdateSeconds         = "since_last_update_seconds"
+	usageTotalBytes                     = "total_bytes"
+	usageObjectsTotal                   = "objects_total"
+	usageVersionsTotal                  = "versions_total"
+	usageDeleteMarkersTotal             = "delete_markers_total"
+	usageBucketsTotal                   = "buckets_total"
+	usageObjectSizeDistribution         = "object_size_distribution"
+	usageObjectVersionCountDistribution = "object_version_count_distribution"
+)
+
+var (
+	usageSinceLastUpdateSecondsMD = NewGaugeMD(usageSinceLastUpdateSeconds,
+		"Time since last update of usage metrics in seconds")
+	usageTotalBytesMD = NewGaugeMD(usageTotalBytes,
+		"Total cluster usage in bytes")
+	usageObjectsTotalMD = NewGaugeMD(usageObjectsTotal,
+		"Total cluster objects count")
+	usageVersionsTotalMD = NewGaugeMD(usageVersionsTotal,
+		"Total cluster object versions (including delete markers) count")
+	usageDeleteMarkersTotalMD = NewGaugeMD(usageDeleteMarkersTotal,
+		"Total cluster delete markers count")
+	usageBucketsTotalMD = NewGaugeMD(usageBucketsTotal,
+		"Total cluster buckets count")
+	usageObjectsSizeDistributionMD = NewGaugeMD(usageObjectSizeDistribution,
+		"Cluster object size distribution", "range")
+	usageObjectVersionsDistributionMD = NewGaugeMD(usageObjectVersionCountDistribution,
+		"Cluster object version count distribution", "range")
+)
+
+// loadClusterUsageObjectMetrics - reads cluster usage metrics.
+//
+// This is a `MetricsLoaderFn`.
+func loadClusterUsageObjectMetrics(ctx context.Context, m MetricValues) error {
+	objLayer := newObjectLayerFn()
+	if objLayer == nil {
+		return nil
+	}
+
+	// Collect cluster level object metrics.
+	dataUsageInfo, err := loadDataUsageFromBackend(ctx, objLayer)
+	if err != nil {
+		logger.LogIf(ctx, err)
+		return nil
```

Comment:
with the break up of metrics info across endpoints, there is redundant calls to load usage/ fetch storage info - perhaps it may be ideal to cache data for related metric endpoints once as opposed to caching per metric endpoint

Reply:
We can start heavily using `cachevalue.Cache` implementations here, we really don't need such frequent loads from disk.

Reply:
This is done now, a separate structure holds all the cached metrics. Not everything is cached.